Dont delete remote elements during mesh generation

lindsayad opened the pull request:
Refs #14451

lindsayad committed:
Dont delete remote elements during mesh generation

Refs #14451

lindsayad commented:
I might be able to demonstrate usefulness of this with mortar, but that also may have to wait on libmesh/libMesh#2354

lindsayad committed:
Simplify and robustify AugmentSparsityOnInterface

Previously we only attempted to ghost lower dimensional
elements and then through them, their interior parents. However,
one could envision a scenario where the interior parents might
be deleted before we even generate the lower D elements. To
avoid that scenario we ghost boundary elements indepent of
whether they are attached to lower d elems

lindsayad committed:
Add lower d elements on all procs

Even in DistributedMesh mode we want to add lower d elements on
all procs otherwise there will be no lower d elements to ghost,
e.g. our mortar interface will be incomplete

lindsayad committed:
Demonstrate some mortar distributed mesh

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/437967/) on 09dbfe2 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14458/site)

This comment will be updated on new commits.

lindsayad committed:
Avoid repartitioning when possible before deleting remote elems

lindsayad commented:
These tests (in distributed mode) will also have to wait on libmesh/libMesh#2354:
```
mesh_modifiers/lower_d_block.second_order ..................................................... FAILED (CRASH)
auxkernels/element_aux_var.high_order_l2_test ................................................. FAILED (CRASH)
mesh/mesh_generation.spiral_annular_mesh ...................................................... FAILED (CRASH)
```

lindsayad committed:
Add geometric ghosting for a node-face problem

Also continue to ensure that remote element deletion doesn't
occur until after proxy relationship managers have been added
for displaced problems. This latter bit actually makes the
FaceFaceRelationshipManager unnecessary because when the
penetration locator is added by the constraint then we inform
the MooseMesh that we need its associated boundaries ghosted.
In fact even if we delete remote all the non-local elements
after equation systems init (e.g. if we don't have any constraints
or materials that request penetration locators before the
equation systems init) the ghost boundary elements that we need
will actually be explicitly added back to the mesh when
MooseMesh::ghostGhostedBoundaries is called during initialSetup.

friedmud reviewed (changes requested):
The reality here is more nuanced than this.  I actually have a use-case where a mesh generator (`FancyExtruderGenerator`) _needs_ to have the remote elements removed _before_ running (only way to generate ~5 Billion element meshes).

The reality here is that Generators need to require a mixture of "traits".  @permcody and I have talked extensively about this.

In addition: we want all generators to work with Distributed mesh _without remote elements_.  That doesn't mean it's always possible (time/money/people constraints)... but it should be the default mode of working.  A Generator should throw an error if a Distributed mesh comes through and it doesn't work with it.

friedmud commented:
As a follow up: we just need to talk about this more.  Let's discuss when you're here @lindsayad

lindsayad committed:
Remove unnecessary face-face relationship manager

Eventually we will probably want to convert our geometric
search stuff over to using ghosting functors, but for now
everything works as-is

lindsayad committed:
Add updated libmesh so we have all_second_order correct

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 143b57032761c71a1df09c59ffc692bb4f198a0c
+Subproject commit dee58a3a6c9c1a69a42e37fa24df214c2a83aa2b
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

lindsayad commented:
> we want all generators to work with Distributed mesh _without remote elements_.

What specifically do you mean by this? I think `LowerDBlockFromSidesetGenerator` should work with `DistributedMesh`, but in order for it to work it would require holding onto the possibly remote higher dimensional boundary elements from which the lower-d elements are generated. My solution for this would be to actually specify a geometric ghosting functor in the `LowerDBlockFromSidesetGenerator::validParams` and then make sure that geometric ghosting functors are added before `MeshGenerators` are executed. I would like this solution better than saying that any input file that uses my generator must be run with `ReplicatedMesh`

permcody commented:
> and then make sure that geometric ghosting functors are added before MeshGenerators are executed.

We definitely need this part for sure regardless of what else we decide.

YaqiWang commented:
We have to make our majority of applications work first. We have seen in the past that some codes failed miserably due to the bad focus.

lindsayad commented:
> We definitely need this part for sure regardless of what else we decide.

Unless we really don't want mesh generators to have access to remote elements.

permcody commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -300,6 +300,12 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   * Used to examine the state of \p _needs_prepare_for_use
+   * @param needs_prepare_for_use Will be set to the current value of \p _needs_prepare_for_use
+   */
+  void needsPrepareForUse(bool & needs_prepare_for_use) const;
```
Ew...

permcody commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -2328,6 +2334,12 @@ MooseMesh::needsPrepareForUse()
+void
+MooseMesh::needsPrepareForUse(bool & needs_prepare_for_use) const
+{
+  needs_prepare_for_use = _needs_prepare_for_use;
```
Oh, this is just a getter and the other method is a pure setter. Can we clean this up rather than having the getting return through a parameter and the setter be parameterless?

lindsayad replied:
If any of the apps use this API I volunteer you to clean it up :smile:

permcody commented on framework/src/meshgenerators/LowerDBlockFromSidesetGenerator.C:
```diff
@@ -143,7 +137,10 @@ LowerDBlockFromSidesetGenerator::generate()
-      side_elem->set_id(max_elem_id + processor_id() * max_elems_to_add + i);
+    {
+      side_elem->set_id(max_elem_id + i);
+      side_elem->set_unique_id() = max_elem_id + i;
```
How can this be unique in parallel?

permcody replied:
Oh - because we aren't deleting remotes yet?

lindsayad replied:
Right, and even if we do delete *some* remotes, I would have a geometric ghosting functor that ensures that all the boundary elements were available on all processes and that we would be generating these lower dimensional elements on all processes, e.g. they would be owned by one process and ghosts on all other processes. Hence since we are adding on all processes we actually want their IDs to be the same.

permcody commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -2076,6 +2076,12 @@ MooseMesh::buildMeshBaseObject(ParallelType override_type)
+  // chance to add geometric ghosting functors. (You can look at our task list in Moose.C to see
+  // that we execute mesh generators before addition of geometric ghosting). We need to delay remote
+  // element deletion until after we've had a chance to add those functors
+  mesh->allow_remote_element_removal(false);
```
@friedmud was concerned about not having the ability to control this. Should we have a parameter for saying, no - really delete for advanced usage?

lindsayad replied:
It sounds like @friedmud would prefer to have the default be to allow remote element removal, and I don't necessarily disagree. When I put this up, it was for like a few week stop-gap until we arrived at a more mature design idea. I wouldn't be scarred if we decided not to move forward with this PR although it seems like there are some who would like it if we did, e.g. @YaqiWang @arovinelli.

My preferred design is to always allow remote element deletion but as I've said allow `MeshGenerators` do add ghosting functors.

permcody replied:
> My preferred design is to always allow remote element deletion but as I've said allow MeshGenerators to add ghosting functors.

Agreed - this is the right fix even if it's not immediately obvious that we need it.

lindsayad replied:
So maybe I'll head in that direction with this PR, or should I wait until the 16th?

arovinelli commented:
@lindsayad I’m perfectly fine with allowing ghosting at RM to work with mesh generators. The only suggestion I have, if this is the selected route is to improve RM documentation, set in place some RM si he that basic ghosting functionality are covered (there are a few) maybe it’s enough to advertise them more) and ar least the most common users mesh generators (e.g sidesetbynormals’ to work in distributed mode. Otherwise it will be useless

Sent from my iPhone

On Dec 3, 2019, at 6:07 PM, Alex Lindsay <notifications@github.com> wrote:

﻿

@lindsayad commented on this pull request.

________________________________

In framework/src/mesh/MooseMesh.C<https://github.com/idaholab/moose/pull/14458#discussion_r353488339>:

> @@ -2076,6 +2076,12 @@ MooseMesh::buildMeshBaseObject(ParallelType override_type)
   if (!getParam<bool>("allow_renumbering"))
     mesh->allow_renumbering(false);

+  // MeshGenerators may call `prepare_for_use` which would delete remote elements before we've had a
+  // chance to add geometric ghosting functors. (You can look at our task list in Moose.C to see
+  // that we execute mesh generators before addition of geometric ghosting). We need to delay remote
+  // element deletion until after we've had a chance to add those functors
+  mesh->allow_remote_element_removal(false);


It sounds like @friedmud<https://github.com/friedmud> would prefer to have the default be to allow remote element removal, and I don't necessarily disagree. When I put this up, it was for like a few week stop-gap until we arrived at a more mature design idea. I wouldn't be scarred if we decided not to move forward with this PR although it seems like there are some who would like it if we did, e.g. @YaqiWang<https://github.com/YaqiWang> @arovinelli<https://github.com/arovinelli>.

My preferred design is to always allow remote element deletion but as I've said allow MeshGenerators do add ghosting functors.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/pull/14458?email_source=notifications&email_token=AIIMUU3S6WARSXFYSM2QIC3QW3YBPA5CNFSM4JSMTLI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCN3HL4Q#discussion_r353488339>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AIIMUU2R27PCLDB3HSK7DELQW3YBPANCNFSM4JSMTLIQ>.

lindsayad commented:
Closing this because our default recommendation is that `MeshGenerators` should work with as few non-local elements as possible. This is the total opposite direction and we don't want to down it. We want users to do combinations of `RelationshipManagers` with `MeshGenerators`, that will work on `DistributedMeshes`, that communicate what elements need to be ghosted in order for their `MeshGenerator` to work, assuming we still have the ghost elements available (think not pre-split). If we don't have ghost elements available, e.g. we have pre-split, then we want users to be get comfortable using the `push` and `pull` APIs from libMesh.

arovinelli commented:
>  We want users to do combinations of `RelationshipManagers` with `MeshGenerators`, that will work on `DistributedMeshes`, that communicate what elements need to be ghosted in order for their `MeshGenerator` to work, assuming we still have the ghost elements available (think not pre-split). If we don't have ghost elements available, e.g. we have pre-split, then we want users to be get comfortable using the `push` and `pull` APIs from libMesh.

Within this regard I believe that #14456 needs to use a relationship manager inside the mesh generator. We have managed to make all tests pass for the distributed mesh, even in the debug mode, but when I apply an interface kernel on the generate boundary it is like if the neighbor is lost (it always give zero variable and material property values)

lindsayad commented:
> Within this regard I believe that #14456 needs to use a relationship manager inside the mesh generator. We have managed to make all tests pass for the distributed mesh, even in the debug mode, but when I apply an interface kernel on the generate boundary it is like if the neighbor is lost (it always give zero variable and material property values

We probably do need a relationship manager there. I'm a little surprised though that the issue is zero values; I would expect that if the neighbor is totally missing then we would be getting failed assertions in libMesh or an error out of PETSc about attempts to index into a vector when the process doesn't hold that index, e.g. it hasn't been ghosted

arovinelli commented:
@lindsayad  I spent some time last night on it and in fact is failing an assertion in debug mode (see [here](https://github.com/idaholab/moose/pull/14456#issuecomment-567050384)) but I'm not sure what does that mean.

permcody commented:
We only test assertions in devel and debug mode. They are optimized out in
the remaining configuration modes. This is normal assertion behavior for
C/C++ and meant to encourage prolific use without optimization concerns.

On Wed, Dec 18, 2019 at 9:43 AM Andrea Rovinelli <notifications@github.com>
wrote:

> @lindsayad <https://github.com/lindsayad> I spent some time last night on
> it and in fact is failing an assertion in debug mode (see here
> <https://github.com/idaholab/moose/pull/14456#issuecomment-567050384>)
> but I'm not sure what does that mean.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/14458?email_source=notifications&email_token=AAXFOIDTJMKVQGHKAL5D6CDQZJHLXA5CNFSM4JSMTLI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHGXSSA#issuecomment-567114056>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXFOIA35GUHMVL3W7LUYH3QZJHLXANCNFSM4JSMTLIQ>
> .
>
